crates/pallet-native-to-evm-swap/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+//! A substrate pallet containing the native to evm currency swap integration.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::fungible::Inspect;
+pub use pallet::*;
+pub use weights::*;
+
+pub mod weights;
+
+/// Utility alias for easy access to the [`Inspect::Balance`] of the [`Config::NativeCurrency`] type.
+type NativeBalanceOf<T> =
+    <<T as Config>::NativeCurrency as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// Utility alias for easy access to the [`Inspect::Balance`] of the [`Config::EvmCurrency`] type.
+type EvmBalanceOf<T> =
+    <<T as Config>::EvmCurrency as Inspect<<T as Config>::EvmAccountId>>::Balance;
+
+// We have to temporarily allow some clippy lints. Later on we'll send patches to substrate to
+// fix them at their end.
+#[allow(clippy::missing_docs_in_private_items)]
+#[frame_support::pallet]
+pub mod pallet {
+    use fp_ethereum::ValidatedTransaction;
+    use frame_support::{
+        pallet_prelude::*,
+        sp_runtime::traits::{Convert, MaybeDisplay, UniqueSaturatedInto},
+        sp_std::fmt::Debug,
+        storage::with_storage_layer,
+        traits::{
+            fungible::Mutate,
+            tokens::{Preservation, Provenance},
+        },
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_core::{H160, H256, U256};
+
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    /// Configuration trait of this pallet.
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_ethereum::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The EVM user account identifier type.
+        type EvmAccountId: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Debug
+            + MaybeDisplay
+            + Ord
+            + MaxEncodedLen
+            + Into<H160>;
+
+        /// Native currency.
+        type NativeCurrency: Inspect<Self::AccountId> + Mutate<Self::AccountId>;
+
+        /// EVM currency.
+        type EvmCurrency: Inspect<Self::EvmAccountId>;
```

Comment:
```suggestion
        /// Native currency.
        type NativeToken: Inspect<Self::AccountId> + Mutate<Self::AccountId>;

        /// EVM currency.
        type EvmToken: Inspect<Self::EvmAccountId>;
```

In the interest of moving away from currency - I think this naming change makes sense. Or should we use `Fungible`, or keep as-is instead? Let's think about it?

Reply:
Good suggestion. In my opinion, token option sounds better as we usually talk about `HMND` token and `eHMND` token. One more point that we are moving away from currency.

Reply:
https://github.com/humanode-network/humanode/pull/1449/commits/9ffac3469a27e3614780933659c4c1fa78179e5b